undofx/src/main/java/org/fxmisc/undo/impl/UndoManagerImpl.java

```diff
@@ -209,7 +229,7 @@ private void addChange(C change) {
     @SuppressWarnings("unchecked")
     private C[] merge(C c1, C c2) {
         Optional<C> merged = merge.apply(c1, c2);
-        if(merged.isPresent()) {
+        if(merged.isPresent() && isNotIdentity.test(merged.get())) {
             return (C[]) new Object[] { merged.get() };
         } else {
             return (C[]) new Object[] { c1, c2 };
```

Comment:
Now that you inlined it, I see one problem, though. If the changes can annihilate, it can cause merging of changes that were previously prevented from being merged.

When this is the current queue

```
... c1
```

and `preventMerge` is invoked, there will be a _temporary_ barrier for merging `c1` (`canMerge` set to `false`):

```
... c1|
```

Then when `c2` arrives, merge will not even be attempted, but the barrier is erased.

```
... c1 c2
```

Then if `c3` = <code>c2<sup>-1</sup></code> is received, `c2` is annihilated and `c1` is no longer prevented from being merged.

```
... c1
```